### ISSUE TITLE
Support POST for authorization code request flow

### ISSUE THREAD

**Issue by sylvain-costanzo** [spring-projects/spring-authorization-server#1811]
**Expected Behavior**
A POST /authorize without the `openid` scope should be accepted by the authorization server

**Current Behavior**
A POST /authorize without the openid scope returns a 404 with the following error log :
```
org.springframework.web.servlet.resource.NoResourceFoundException: No static resource oauth/authorize
```
![image](https://github.com/user-attachments/assets/6d4a5523-41ea-462a-b78f-e870d6b15b4f)

**Context**
Our implementation allow to not pass the scope parameter during the first /authorize request, as allowed by the oauth2 specification
https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.1 and https://datatracker.ietf.org/doc/html/rfc6749#section-3.3
In this case our server act as if every scopes were requested.

There is no issue with spring authorization server when using a GET, but our implementation replays the /authorize request with a POST after a /login.

The `authorizationEndpointMatcher` of `OAuth2AuthorizationEndpointFilter` only match POST /authorize requests who have the `openid` scope : https://github.com/spring-projects/spring-authorization-server/blob/main/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java#L156

I guess this is because the oauth2 spec states that the authorization server MAY support the use of POST https://datatracker.ietf.org/doc/html/rfc6749#section-3.1
while openid spec states that the server MUST support GET and POST https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

Unfortunately we cannot submit this second /authorize with a GET as it would break other parts of our system.
We considered tinkering with the RequestCache to add every scopes of the client in the SavedRequest, but a modification in spring authorization server would be simpler because we can just handle this use case in the token generators.

Is it possible to allow the customization of this matcher, or to remove the `openid` scope matcher for POST in OAuth2AuthorizationEndpointFilter ?

**Comment by sylvain-costanzo** [comment]
Hello @jgrandja  and thank you for your answer

The main issue comes from our custom implementation of the `prompt=login`.
We are using an authentication converter to catch the `/authorize` request if `prompt=login` is detected within a GET request.
Using a POST during our "replay" of the `/authorize` request allow us to not loop indefinitely because of this converter.

We understand that our implementation may be sub-optimal, and that it is the first time someone request this feature.
But as the oauth2 spec states that it MAY support POST, we thought it was not an unreasonable request to do.

**Comment by jgrandja** [comment]
@sylvain-costanzo I don't think we want to expose a setter for `authorizationEndpointMatcher` but we will consider adding support for `POST` for a standard OAuth 2.1 Authorization Request flow. I've changed the title of the ticket to reflect that enhancement.

**Comment by sylvain-costanzo** [comment]
@jgrandja 
I am willing to work on this feature but I don't see it planned in [this board](https://github.com/orgs/spring-projects/projects/8/views/21) so I am reaching out to know if it would be acceptable ?

I think the best way to do it is to remove the `openid` scope matcher in [OAuth2AuthorizationEndpointFilter](https://github.com/spring-projects/spring-authorization-server/blob/main/oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2AuthorizationEndpointFilter.java#L156) so that's how I would implement it.

**Comment by jgrandja** [comment]
@sylvain-costanzo A PR for this enhancement would be greatly appreciated.

> I think the best way to do it is to remove the `openid` scope matcher

I think that will work.
